### DISCUSSION TITLE
Run mutation on page load

### DISCUSSION THREAD

**Discussion by devuxer** [TanStack/query#986] [General]
I want to have a mutation run when a page loads. Would you recommend putting a `mutate()` call inside a `useEffect` in this case or is there a better way?

My use case is a user clicking a link in a transactional email to verify their email account.

**Comment by justinthareja** [comment]
@devuxer how did you end up solving this?

I've recently run into this issue myself. I'm still new to the react ecosystem, and for various reasons my click handler needs to be wrapped in a `useCallback`. So right now the code looks like:

```js
// custom react-query hooks as per recommendation
function usePlay() {
  const { user } = useUser();
  const mutation = useMutation((options) => spotifyApi.play(options));
  return user ? mutation : stub;
}

function usePause() {
  const { user } = useUser();
  const mutation = useMutation(() => spotifyApi.pause());
  return user ? mutation : stub;
}

// consuming custom hooks
const pause = usePause();
const play = usePlay();

const playPauseHandler = useCallback(() => {
    if (status.state === "playing") {
      pause.mutate();
    } else {
      play.mutate({
        uris: [`spotify:track:${item.id}`],
        position_ms: progress,
      });
    }
  }, [item.id, status.state, progress]); `` // how can i properly include mutate here?
```

I'm still trying to wrap my head around hooks and react-query. Is it ok to omit the `mutate` function from the dependency array? If I include it, I believe there is an infinite re-render cycle because `mutate !== mutate` for each cycle. It seems to work as expected if I exclude it.

From my limited understanding I could somehow properly include `mutate` in the dep array by wrapping it in it's own useCallback somehow? Still a little fuzzy on this, please advise.

**Reply by devuxer** [reply]
@justinthareja, I just went with putting the mutation call inside a `useEffect`. Seems to work.

Here's an example:

```ts
const linkToken = useUrlSearchParam("token"); // extracts token from URL
const { mutateAsync, isLoading, error } = usePost<Payload>("/account/apply-changed-email-address");
const { logOut } = useSession();
useEffect(() => {
    void applyChangedEmailAddress();

    async function applyChangedEmailAddress() {
        if (!linkToken) return;
        const payload: LinkTokenPayload = { linkToken };
        await mutateAsync(payload);
        if (!error) void logOut();
    }
}, [linkToken, mutateAsync, logOut]);
```

No idea if this implementation is ideal, but the mutation does get called.

**Reply by IanVS** [reply]
Be careful if you're triggering a mutation in `useEffect` in React 18, it will fire twice in dev with strict mode, which might not be what you want.

**Reply by ganesansundareswaran** [reply]
@IanVS any solution for this?

**Reply by IanVS** [reply]
This is a late response, and I'm not sure it's a good one, but thus far I've used a ref to avoid the second mutation call. 

```ts
  const hasFired = useRef(false);
  const {mutate, isIdle} = useMutation(/* ... */);

  useEffect(() => {
    // Only trigger mutation once, even if useEffect fires more times
    if (isIdle && !hasFired.current) {
      hasFired.current = true;
      mutate();
    }
  }, [isIdle, mutate]);

```

**Reply by reorx** [reply]
@IanVS This trick works for restricing mutate to be called only once, but I found that the states returned by useMutation would not update after it's done, do you find the same problem?

**Reply by IanVS** [reply]
Interesting.  I see the status go from idle to loading, and then stay stuck there.  I wasn't using the status so I didn't notice, but yeah that seems to be a problem.  Curious if anyone else has a solution.  I'm sure @TkDodo has answered it on Twitter a hundred times.

**Reply by TkDodo** [reply]
I think I discussed this somewhere with a codesandbox reproduction, not sure where. The "problem" is that `useMutation` doesn't share state between runs. You can call `useMutation` twice in two components, and it will have different data. In fact, every time you call `.mutate()`, the data that is tracked by `useMutation` is tied to the latest invocation.

So in this case, it's really the ref workaround that introduces the issue. In the first run, you call `useMutation`, trigger `mutate`, and then `useMutation` tracks that. Then, strict mode kicks in, and we call `useMutation` again, but not `mutate`. At this point, we're no longer having an observer for the initial mutate call, so the `useMutation` observers something different now.

In v5, you could call `useMutationState` to work around this, but I really wouldn't build ref workarounds just to stop something in dev mode. If you are concerned about future offscreen apis from react running your effects multiple times, than you really shouldn't put mutations inside effects. Mutations are not a side-effect of a component mounting / rendering. They should be side effects of user interactions. The initial comment states:

> I want to have a mutation run when a page loads.

And I'm always questioning that statement per se. Why is that? If the page loads and you want to do something, do it on the server. Or, do it in the user interaction that got the user to that page. If you don't have a server (SPA) _and_ users can go to this page directly, then yeah, idk, maybe the effect is fine. But then it also should be fine if it runs twice, because users can just reload the browser page a bunch of times, in which case, it's also fired multiple times.

**Reply by reorx** [reply]
> Mutations are not a side-effect of a component mounting / rendering. They should be side effects of user interactions.

I 100% agree with this statement, so I decided to solve the problem in a different way.

My problem arose from modularizing and refactoring the code. Specifically, I was building a modal component for product checkout. Initially, when a user clicked a button, the onClick function would first open the modal and then call the `.mutate` method with the specified product ID. To reduce redundancy in the code, I encapsulated the logic in the CheckoutModal component and used the useEffect function to call `.mutate` when isOpen was true. This made the component the one that needed to have a mutation run when it was rendered.

Theoretically, the mutation should be triggered by a user click, but in practice, it became a side effect of a component. To avoid the problem that `useMutation` doesn't share state between runs, I simply relocated `useMutation` outside of the component, which failed my intended goal of refactoring. However, there must be a better solution that doesn't defy the design philosophy of mutation. If I find one in the future, I will update this thread.

**Reply by IanVS** [reply]
In general I also agree, but I've got a case that requiring user interaction would be a bit of a bummer.  I make a `POST` as part of a magic-link registration flow.  The user clicks a link in their email, they're taken to a page of our SPA, and they need to be given an authenticator code to save in their 2fa app and verify.  The call to create the authenticator code is a POST and needs to happen only once to avoid race conditions.  Yes, we could just accept that it's broken in dev, but I'd really rather not.  The ref hack works fine for me, since it only impacts dev, and I don't actually use the status of the mutation.

**Reply by TkDodo** [reply]
Sounds like it might also be broken if the user clicks the link twice?

if you don't use the status, you can also make the request without useMutation

**Reply by IanVS** [reply]
>Sounds like it might also be broken if the user clicks the link twice?

I don't think so, since each request would be made in its own tab/window.

>if you don't use the status, you can also make the request without useMutation

Yes, but I'd still have to make sure it only happens once, so I don't think I'd gain anything by avoiding useMutation.  Might as well keep it consistent with the rest of the app.

**Reply by reorx** [reply]
@IanVS Sound similar to a newsletter unsubscription process. Users would appreciate not having to click the button again. Very reasonable, now I'm not 100% sure about the philosophy that "Mutations should be side-effects of user interactions" 😂

**Reply by TkDodo** [reply]
> Sound similar to a newsletter unsubscription process.

I mean yes, but why does this need client side mutation at all? If I go to the `yourservice.com/unsubscribe` page, I would expect to be unsubscribed immediately (that means: the server does it, right away).

having this page be a client side page that needs javascript to actually perform the unsubscribe with a client side request on-mount seems wrong. I guess this is one of the reasons why react tries to push us towards frameworks like next/remix. None of this needs to happen on the client if we have a server.

**Reply by reorx** [reply]
@TkDodo I have had the same question for many years, and I have always hated it. But one day, I realized that the reason behind this phenomenon is likely to avoid unintentional unsubscribing due to anti-spam mechanisms. Some email clients may visit the links in the email to check if they're unsafe. Of course, this case does not apply to those who deliberately make the unsubscribe process difficult for users.

@IanVS Are you not generating the authentication code on the server side for the same reason?

**Reply by IanVS** [reply]
@reorx this seems a bit off-topic for this thread, but yes the code is generated by the server, in response to a POST request.

**Reply by mrclrchtr** [reply]
As @TkDodo advised, one can use `mutationState`.

But I agree overall that it should be done on the server side. 

Nevertheless, here is the solution with `mutationState`:

```typescript
const hasFired = useRef(false);
const confirmSignUp = useConfirmSignUpMutation();

const mutationState = useMutationState({
    filters: { mutationKey: [MUTATION_KEY_CONFIRM_SIGN_UP] },
    select: (mutation) => mutation.state.status,
});

useEffect(() => {
    // Only trigger mutation once, even if useEffect fires more times
    if (confirmSignUp.isIdle && !hasFired.current) {
      hasFired.current = true;
      confirmSignUp.mutate(token);
    }
}, [token, confirmSignUp]);

....

return (
    <Container size={420} my={40}>
      {(() => {
        switch (mutationState[0]) {
          case "idle":
            return confirmSignUpPending;
          case "pending":
            return confirmSignUpPending;
          case "success":
            return confirmSignUpSuccessful;
          case "error":
            return confirmSignUpError;
        }
      })()}
    </Container>
```

**Reply by AnasSiddiqui18** [reply]
Thank you very much. You saved my day 👍

**Comment by TkDodo** [comment]
putting `mutate` or `mutateAsync` in a `useEffect` is fine. Just be prepared that the mutation will run whenever the component mounts. Not sure what the use-case is for that, but there might be some :)

The `mutate` and `mutateAsync` functions should be stable across re-renders so there is no need to omit them from the dependency list of the effect (which is something you generally shouldn't be doing imo)

> My use case is a user clicking a link in a transactional email to verify their email account.

I don't really see why for this use-case, you need to make a request from the frontend:
- user requests page, token or so is probably in the url
- page loads
- you make a mutation onMount

I would just do:
- user requests page, token or so is probably in the url
- backend reads the url, validates the email
- delivers page to frontend, user is validated or not 🤷‍♂️

**Reply by justinthareja** [reply]
Thanks for that input @TkDodo. I dug a little deeper to see why `mutate` and `mutateAsync` are stable between re-renders and it looks like they are wrapped in a useCallback before being exported. Helped with my issue at least!

![Screen Shot 2021-03-30 at 1 36 38 PM](https://user-images.githubusercontent.com/8769998/113031818-25f91900-915d-11eb-825d-1aa07bf5367a.png)

**Reply by devuxer** [reply]
> I would just do:
> 
> * user requests page, token or so is probably in the url
> * backend reads the url, validates the email
> * delivers page to frontend, user is validated or not 🤷‍♂️

Thanks for the reply, @TkDodo!

Interesting suggestion, though I'm using Next.js, and I'd honestly have to research how to do what you're suggesting.

**Reply by chiptus** [reply]
@TkDodo am I right to assume that although `mutate` is stable, the result of `useMutation` isn't?

for some reason, eslint shouted at me when I put `useMutationResult.mutate` in the dependency list and it expected me to put the whole result object as a dependency which caused numerous mutations. while doing:
```
const {mutate} = useMutationResult

useEffect(() => {
	mutate()
}, [mutate])
```

worked. any idea why?

**Reply by TkDodo** [reply]
> am I right to assume that although mutate is stable, the result of useMutation isn't?

yes, that's right.

> any idea why?

I have seen that as well, my best guess is that its a shortcoming of the lint rule :(

**Reply by TkDodo** [reply]
i think it's this:
- https://github.com/facebook/react/issues/16265

**Reply by chiptus** [reply]
thanks for the quick answer!

**Reply by chiptus** [reply]
and here's the response https://github.com/facebook/react/issues/16265#issuecomment-517518539 which makes sense

**Comment by vashisth00** [comment]
Any other solution for this? 👎🏻

**Comment by tonypee** [comment]
useEffect seems like the right place, as triggering the mutation is an imperative action, which can be called at any time - on the useEffect is a particular time. You just need to decide if that is the time you want

**Reply by lamualfa** [reply]
If you want to run the mutation just once when using React `StrictMode`, you can do this:

```js
const mutation = useMutation()

useEffect(() => {
   if(mutation.isIdle) {
      mutation.mutate()
   }
}, [])

```

**Comment by yairopro** [comment]
A little bit dirty (since a state is created when not needed) but does the job
```javascript
const mutation = useMutation(...);
useState(mutation.mutate); // run only once per component
```

**Reply by TkDodo** [reply]
Sorry to say but useState initialzers should be pure functions and not have side effects like running a mutation

**Reply by yairopro** [reply]
Sorry to say but `useMutation` should have a way to be run at mount (in strict mode). 😂

**Reply by TkDodo** [reply]
sorry to say, firing a side-effect depending on when a component renders or not doesn't seem to play well with how react envisions side-effects 🤷‍♂️

**Reply by yairopro** [reply]
Hi @TkDodo :)

Maybe I am wrong, but the react strict-mode fires side-effects twice in development in order to help them to be _robust to re-renders and re-mounting_ (forcing us to deduplicate them I guess) before production.
That's why having a `useMutation()` with an option to run it on mount (and deduplicate it with something like a nonce 😇🙈) can then indeed play well with how react envisions side-effects.

**Reply by TkDodo** [reply]
directly from [the react docs](https://react.dev/learn/you-might-not-need-an-effect#recap):

> Code that runs because a component was _displayed_ should be in Effects, the rest should be in events.

That's the reason strict mode runs everything multiple times. There are new APIs coming in React19 that might fire your effects more often in production as well, e.g. when something goes out of screen or becomes visible again after it has been hidden behind a Modal Dialog (used to be called `offscreen` API, now [will be called `Activity`](https://react.dev/blog/2024/02/15/react-labs-what-we-have-been-working-on-february-2024#offscreen-renamed-to-activity)).

So they really want you to think about if your effect is related to a component being rendered / shown, or if you're just doing that because you think it's the best "lifecycle" to hook into. Usually, mutations don't fall into that category. You usually don't want to fire a mutation because a component renders - you want it because the user interacted with something in your app. One exception imo is tracking pageViews to an analytics API. But there it doesn't matter if it fires twice in devmode, as it won't be fired doubled in production (e.g. if you have an effect in your top page component).

**Reply by yairopro** [reply]
THank you  
You rock

**Comment by tonypee** [comment]
Tony Polinelli

**Comment by devuxer** [comment]
> So they really want you to think about if your effect is related to a component being rendered / shown, or if you're just doing that because you think it's the best "lifecycle" to hook into. Usually, mutations don't fall into that category. You usually don't want to fire a mutation because a component renders - you want it because the user interacted with something in your app. One exception imo is tracking pageViews to an analytics API. But there it doesn't matter if it fires twice in devmode, as it won't be fired doubled in production (e.g. if you have an effect in your top page component).

In my case, it's that I want to fire a mutation because the user clicked on a link in a transactional email to, say, reset their password, and I want to fire a mutation due the action of the user clicking that link the email. Since I can't tie into that user action from the email client, I tie it to the page loading in the browser, than I display the appropriate message in the browser (success, link expired, or whatever).

**Reply by TkDodo** [reply]
Usually I would say: Let the server handle this when the specific route is accessed. But even if you do client side only, you can fire a request when the bundle evaluates, outside of the react lifecycle. Or really, just do it in a useEffect in the top level component that you pass to `root.render`:

```
const root = ReactDOM.createRoot(rootElement)
root.render(
  <>
    <React.StrictMode>
      <App /> 
     </React.StrictMode>
     <SomeComponentThatRendersNullAndFiresAMutationInUseEffect />
  </>
)
```

really, an effect there is fine ...

**Reply by devuxer** [reply]
Thanks @TkDodo!

This is a NextJS app (w/ pages folder), and the effect is at the page level, which is the top-level component, so I think I'll just stick with my current `useEffect` solution for now. Helpful to get your take on these alternatives, though!

**Reply by TkDodo** [reply]
if you have a server, why do you need to "fire a mutation because of the action of the user clicking that link in the email" from the browser? You have a server. The user clicks the link, it navigates to `https://my-app/rest-password`. That could be an API route that does what you want the mutation to do, then redirect to a page. Or, it would be a page that does what the mutation would do in `getServerSideProps`...

**Reply by devuxer** [reply]
> if you have a server, why do you need to "fire a mutation because of the action of the user clicking that link in the email" from the browser?

Honestly, that would probably work fine, but at the time I built this feature, I didn't know better (and no one responded to my original post here for half a year) . Fast forward to today, and I'm a little reluctant to mess with it as long as it's working. but I may tackle it as a learning experience. It seems like I'd need to handle the request at the server and then serve one of several different pages (or perhaps a single page with a query string) to provide feedback to the user on whether the operation was successful or not. I'd love to see a relevant example in NextJS to see how this would be done.

**Comment by RodolfoSilva** [comment]
Instead of using `useMutation`, why not use `useQuery` like this:

```ts
useQuery({
  queryKey: ['join-channel', channelId],
  queryFn: async () => {
    await doSomething();
    return null;
  },
  staleTime: Infinity,
});
```
This ensures the initialization logic runs properly with async behavior and benefits from React Query’s built-in retry mechanism if something goes wrong.
I’m using a similar approach in a more unstable environment, like on mobile devices, and it’s been working well.
